src/pages/dashboard/.Map.js

```diff
@@ -0,0 +1,155 @@
+import React, { useState, useMemo, useEffect } from 'react'
+
+import { json } from 'd3-fetch'
+import { nest } from 'd3-collection'
+import { sum } from 'd3-array'
+
+import _Map from './charts/_Map'
+import { path } from './charts/_Map'
+import StatesWithPopulation from './data/_state-populations'
+
+import { formatDate, formatNumber } from './.utils'
+
+import './.Map.scss'
+
+const Map = () => {
+  // holds the date of the displayed day
+  const [currentDate, setCurrentDate] = useState('20200401')
+  // holds the field we are currently viewing
+  const [currentField, setCurrentField] = useState('positive')
+
+  const [useChoropleth, setUseChoropleth] = useState(false)
+
+  const [rawStateData, setRawStateData] = useState(null)
+
+  const getValue = useMemo(
+    () => (d, field = currentField, normalized = false) =>
+      ((d.properties.dailyData[currentDate] &&
+        d.properties.dailyData[currentDate][field]) ||
+        0) / (normalized ? d.properties.population / 1000000 : 1),
+    [currentDate, currentField],
+  )
+
+  const joinedData = useMemo(() => {
+    if (!rawStateData || !path) return null
+    const createMapFromArray = (array, keyField, valueField = null) => {
+      return Object.assign(
+        {},
+        ...array.map(a => ({ [a[keyField]]: valueField ? a[valueField] : a })),
+      )
+    }
+    const groupedByState = nest()
+      .key(d => d.state)
+      .entries(rawStateData)
+    const stateMap = createMapFromArray(groupedByState, 'key', 'values')
+    const joinedFeatures = StatesWithPopulation.features.map(feature => ({
+      ...feature,
+      properties: {
+        ...feature.properties,
+        centroidCoordinates: path.centroid(feature), // should get rid of turf and use d3 for the centroid
+        dailyData: createMapFromArray(
+          stateMap[feature.properties.STUSPS],
+          'date',
+        ),
+      },
+    }))
+    return { ...StatesWithPopulation, features: joinedFeatures }
+  }, [rawStateData])
+
+  const sumTotalTestResults = useMemo(
+    () =>
+      joinedData &&
+      sum(joinedData.features, d => getValue(d, 'totalTestResults')),
+    [joinedData, currentDate],
+  )
+  const sumPositive = useMemo(
+    () => joinedData && sum(joinedData.features, d => getValue(d, 'positive')),
+    [joinedData, currentDate],
+  )
+
+  const sumChoro = useMemo(
+    () => joinedData && sum(joinedData.features, d => getValue(d)),
+    [joinedData, getValue],
+  )
+
+  useEffect(() => {
+    const fetchData = async () => {
+      const stateData = await json('https://covidtracking.com/api/states/daily')
+      setRawStateData(stateData)
```

Comment:
So its cool to update two different states in the same function without trigger a double render pass?

Reply:
that was my presumption, and I just confirmed it to be sure.

`useEffect` runs between renders so calling `setSate` multiple times within it will only trigger a single render after the effect is done.

you can see [here](https://codesandbox.io/s/tender-dijkstra-sw6c9?file=/src/index.js). (one render from before the effect is run and one render after the effect is run no matter how many states I set in the effect).